AnalysisTools/plugins/PATObjectValueEmbedder.cc

```diff
@@ -0,0 +1,154 @@
+#include "DataFormats/PatCandidates/interface/Muon.h"
+#include "DataFormats/PatCandidates/interface/Tau.h"
+#include "DataFormats/PatCandidates/interface/Jet.h"
+#include "DataFormats/PatCandidates/interface/Photon.h"
+#include "DataFormats/PatCandidates/interface/CompositeCandidate.h"
+#include "DataFormats/Common/interface/View.h"
+
+
+template<typename T, typename V>
+class PATObjectValueEmbedder : public edm::stream::EDProducer<>
+{
+
+public:
+  explicit PATObjectValueEmbedder(const edm::ParameterSet& iConfig);
+  virtual ~PATObjectValueEmbedder() {};
+
+private:
+  virtual void produce(edm::Event& iEvent, const edm::EventSetup& iSetup);
+  void embedValue(T& object, const int& value, const std::string& label) const;
+  void embedValue(T& object, const bool& value, const std::string& label) const;
+  void embedValue(T& object, const double& value, const std::string& label) const;
+  void embedValue(T& object, const float& value, const std::string& label) const;
+
+  const edm::EDGetTokenT<edm::View<T> > srcToken_;
+  const edm::EDGetTokenT<V> valToken_;
+  const std::string label_;
+};
+
+
+template<typename T, typename V>
+PATObjectValueEmbedder<T,V>::PATObjectValueEmbedder(const edm::ParameterSet& iConfig) :
+  srcToken_(consumes<edm::View<T> >(iConfig.getParameter<edm::InputTag>("src"))),
+  valToken_(consumes<V>(iConfig.getParameter<edm::InputTag>("valueSrc"))),
+  label_(iConfig.getParameter<std::string>("label"))
+{
+  produces<std::vector<T> >();
+}
+
+
+template<typename T, typename V>
+void PATObjectValueEmbedder<T,V>::produce(edm::Event& iEvent,
+                                         const edm::EventSetup& iSetup)
+{
+  edm::Handle<edm::View<T> > in;
+  std::unique_ptr<std::vector<T> > out(new std::vector<T>);
+  edm::Handle<V> value;
+
+  iEvent.getByToken(srcToken_, in);
+  iEvent.getByToken(valToken_, value);
+
+  for(size_t i = 0; i < in->size(); ++i)
+    {
+      edm::Ptr<T> t = in->ptrAt(i);
+
+      out->push_back(*t);
+      embedValue(out->back(), *value, label_);
+    }
+
+  iEvent.put(std::move(out));
+}
+
+template<typename T, typename V>
+void PATObjectValueEmbedder<T,V>::embedValue(T& object,
+                                    const double& value, const std::string& label) const
+{
+    object.addUserFloat(label, value);
+}
+
+template<typename T, typename V>
+void PATObjectValueEmbedder<T,V>::embedValue(T& object,
+                                    const float& value, const std::string& label) const
+{
+    object.addUserFloat(label, value);
+}
+
+template<typename T, typename V>
+void PATObjectValueEmbedder<T,V>::embedValue(T& object,
+                                    const int& value, const std::string& label) const
+{
+    object.addUserInt(label, value);
+}
+
+template<typename T, typename V>
+void PATObjectValueEmbedder<T,V>::embedValue(T& object,
+                                    const bool& value, const std::string& label) const
+{
+    object.addUserInt(label, value);
```

Comment:
Is there an advantage to using `userInt` instead of `userFloat` for bools? I've typically done `userFloat` in the past, and that's what's in a lot of other UWVV plugins, but that may have been a relic of the FSA days. I don't care which we do, but we should probably be consistent.